File: provider/pkg/provider/remoteFile.go
// Copyright 2016-2021, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package provider

import (
	"github.com/pulumi/pulumi-command/sdk/go/command/remote"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// The RemoteFileArgs for creating a RemoteFile component resource.
type RemoteFileArgs struct {
	Connection remote.ConnectionArgs `pulumi:"connection"`
	Content    pulumi.StringInput    `pulumi:"content"`
	Path       pulumi.StringInput    `pulumi:"path"`
}

// The RemoteFile component resource.
type RemoteFile struct {
	pulumi.ResourceState

	Command *remote.Command `pulumi:"command"`
}

// NewRemoteFile creates a new RemoteFile component resource.
func NewRemoteFile(ctx *pulumi.Context,
	name string, args *RemoteFileArgs, opts ...pulumi.ResourceOption) (*RemoteFile, error) {
	if args == nil {
		args = &RemoteFileArgs{}
	}

	component := &RemoteFile{}
	err := ctx.RegisterComponentResource("kubernetes-the-hard-way:index:RemoteFile", name, component, opts...)
	if err != nil {
		return nil, err
	}

	command, err := remote.NewCommand(ctx, name, &remote.CommandArgs{
		Connection: args.Connection,
		Stdin:      args.Content,
		Create:     pulumi.Sprintf("tee %s", args.Path),
		Delete:     pulumi.Sprintf("rm %s", args.Path),
	}, pulumi.Parent(component))
	if err != nil {
		return nil, err
	}

	component.Command = command

	if err := ctx.RegisterResourceOutputs(component, pulumi.Map{
		"command": command,
	}); err != nil {
		return nil, err
	}

	return component, nil
}

File: provider/cmd/pulumi-resource-kubernetes-the-hard-way/remoteFile.ts
import { ComponentResource, ComponentResourceOptions, Input, Inputs, Output, interpolate, output } from '@pulumi/pulumi';
import { ConstructResult } from '@pulumi/pulumi/provider';
import { Command } from '@pulumi/command/remote';
import { remote } from '@pulumi/command/types/input';

export type InstallArgs = Omit<RemoteFileArgs, 'content'> & {
  name: string;
  opts?: ComponentResourceOptions;
};

export interface RemoteFileArgs {
  connection: Input<remote.ConnectionArgs>;
  path: Input<string>;
  content: Input<string>;
}

export class RemoteFile extends ComponentResource {
  public readonly command: Command;
  public readonly content: Output<string>;
  public readonly path: Output<string>;
  public readonly stderr: Output<string>;
  public readonly stdin: Output<string | undefined>;
  public readonly stdout: Output<string>;

  constructor(name: string, args: RemoteFileArgs, opts?: ComponentResourceOptions) {
    super('kubernetes-the-hard-way:index:RemoteFile', name, args, opts);

    const content = output(args.content);
    const path = output(args.path);

    const command = new Command(name, {
      connection: args.connection,
      stdin: content,
      create: interpolate`tee ${path}`,
      delete: interpolate`rm ${path}`,
    }, { parent: this });

    this.command = command;
    this.content = content;
    this.path = path;
    this.stderr = command.stderr;
    this.stdin = command.stdin;
    this.stdout = command.stdout;

    this.registerOutputs({
      command, content, path,
      stderr: command.stderr,
      stdin: command.stdin,
      stdout: command.stdout,
    });
  }
}

export function install({ name, opts, ...rest }: InstallArgs, content: Input<string>): RemoteFile {
  return new RemoteFile(name, {
    connection: rest.connection,
    path: rest.path,
    content,
  }, opts);
}

export async function construct(
  name: string,
  inputs: Inputs,
  options: ComponentResourceOptions,
): Promise<ConstructResult> {
  const file = new RemoteFile(name, inputs as RemoteFileArgs, options);
  return {
    urn: file.urn,
    state: {
      command: file.command,
      content: file.content,
      path: file.path,
      stderr: file.stderr,
      stdin: file.stdin,
      stdout: file.stdout,
    },
  };
}
